src/SAWScript/Prover/Exporter.hs

```diff
@@ -154,55 +170,104 @@ writeSAIGInferLatches proxy sc file tt = do
+  writeCNF proxy sc f t
 
 -- | Write a proposition to an SMT-Lib version 2 file. Because @Prop@ is
 -- assumed to have universally quantified variables, it will be negated.
-writeSMTLib2 :: SharedContext -> FilePath -> Prop -> IO ()
+writeSMTLib2 :: SharedContext -> FilePath -> Prop -> TopLevel ()
 writeSMTLib2 sc f t = writeUnintSMTLib2 mempty sc f t
 
+-- | Write a proposition to an SMT-Lib version 2 file. Because @Prop@ is
+-- assumed to have universally quantified variables, it will be negated.
+-- This version uses What4 instead of SBV.
+writeSMTLib2What4 :: SharedContext -> FilePath -> Prop -> TopLevel ()
+writeSMTLib2What4 sc f t = writeUnintSMTLib2What4 mempty sc f t
+
 -- | Write a @Term@ representing a predicate (i.e. a monomorphic
 -- function returning a boolean) to an SMT-Lib version 2 file. The goal
 -- is to pass the term through as directly as possible, so we interpret
 -- it as an existential.
-write_smtlib2 :: SharedContext -> FilePath -> TypedTerm -> IO ()
+write_smtlib2 :: SharedContext -> FilePath -> TypedTerm -> TopLevel ()
 write_smtlib2 sc f (TypedTerm schema t) = do
-  checkBooleanSchema schema
-  p <- predicateToProp sc Existential [] t
+  io $ checkBooleanSchema schema
+  p <- io $ predicateToProp sc Existential [] t
   writeSMTLib2 sc f p
 
+-- | Write a @Term@ representing a predicate (i.e. a monomorphic
+-- function returning a boolean) to an SMT-Lib version 2 file. The goal
+-- is to pass the term through as directly as possible, so we interpret
+-- it as an existential. This version uses What4 instead of SBV.
+write_smtlib2_w4 :: SharedContext -> FilePath -> TypedTerm -> TopLevel ()
+write_smtlib2_w4 sc f (TypedTerm schema t) = do
+  io $ checkBooleanSchema schema
+  p <- io $ predicateToProp sc Existential [] t
+  writeUnintSMTLib2What4 mempty sc f p
+
 -- | Write a proposition to an SMT-Lib version 2 file, treating some
 -- constants as uninterpreted. Because @Prop@ is assumed to have
 -- universally quantified variables, it will be negated.
-writeUnintSMTLib2 :: Set VarIndex -> SharedContext -> FilePath -> Prop -> IO ()
-writeUnintSMTLib2 unintSet sc f p =
+writeUnintSMTLib2 :: Set VarIndex -> SharedContext -> FilePath -> Prop -> TopLevel ()
+writeUnintSMTLib2 unintSet sc f p = io $
   do (_, _, l) <- prepNegatedSBV sc unintSet p
      let isSat = True -- l is encoded as an existential formula
      txt <- SBV.generateSMTBenchmark isSat l
      writeFile f txt
 
-writeCore :: FilePath -> Term -> IO ()
-writeCore path t = writeFile path (scWriteExternal t)
-
-writeCoreProp :: FilePath -> Prop -> IO ()
-writeCoreProp path (Prop t) = writeFile path (scWriteExternal t)
+-- | Write a proposition to an SMT-Lib version 2 file, treating some
+-- constants as uninterpreted. Because @Prop@ is assumed to have
+-- universally quantified variables, it will be negated. This version
+-- uses What4 instead of SBV.
+writeUnintSMTLib2What4 :: Set VarIndex -> SharedContext -> FilePath -> Prop -> TopLevel ()
```

Comment:
we already have `offline_w4_unint_z3`/`offline_w4_unint_cvc4`/`offline_w4_unint_yices`, let's not add more stuff if it's not necessary.

Reply:
IIUC, those commands generate input that's either somewhat or very specific to those provers. Generic SMT-Lib is also useful.